x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java

```diff
@@ -0,0 +1,554 @@
+                    if (currentStepKey.getAction().equals(AllocateAction.NAME)) {
+                        // this index is in the middle of executing the allocate action - which doesn't exist in the updated policy
+                        // anymore so let's try to move the index to the next action
+
+                        LifecycleExecutionState newLifecycleState = moveStateToNextActionAndUpdateCachedPhase(indexMetadata,
+                            currentExState, System::currentTimeMillis, oldPolicy, newPolicyMetadata, client);
+                        if (currentExState.equals(newLifecycleState) == false) {
+                            mb.put(IndexMetadata.builder(indexMetadata).putCustom(ILM_CUSTOM_METADATA_KEY, newLifecycleState.asMap()));
+                        }
+                    } else {
+                        // if the index is not in the allocate action, we're going to perform a cached phase update (which is "unsafe" by
+                        // the rules defined in {@link PhaseCacheManagement#isIndexPhaseDefinitionUpdatable} but in our case it is safe
+                        // as the migration would've only removed the allocate action and the current index is not in the middle of
+                        // executing the allocate action, we made sure of that)
+
+                        LifecycleExecutionState.Builder updatedState = LifecycleExecutionState.builder(currentExState);
+                        PhaseExecutionInfo phaseExecutionInfo = new PhaseExecutionInfo(newPolicyMetadata.getPolicy().getName(),
+                            newPolicyMetadata.getPolicy().getPhases().get(currentStepKey.getPhase()), newPolicyMetadata.getVersion(),
+                            newPolicyMetadata.getModifiedDate());
+                        String newPhaseDefinition = Strings.toString(phaseExecutionInfo, false, false);
+                        updatedState.setPhaseDefinition(newPhaseDefinition);
+
+                        logger.debug("updating the cached phase definition for index [{}], current step [{}] in policy " +
+                            "[{}] to [{}]", indexMetadata.getIndex().getName(), currentStepKey, policyName, newPhaseDefinition);
+                        mb.put(IndexMetadata.builder(indexMetadata)
+                            .putCustom(ILM_CUSTOM_METADATA_KEY, updatedState.build().asMap()));
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Transition the managed index to the first step of the next action in the current phase and update the cached phase definition for
+     * the index to reflect the migrated phase definition.
+     *
+     * Returns the same {@link LifecycleExecutionState} if the transition is not possible or the new execution state otherwise.
+     */
+    private static LifecycleExecutionState moveStateToNextActionAndUpdateCachedPhase(IndexMetadata indexMetadata,
```

Comment:
Super minor, but I think this method would be better if moved into `IndexLifecycleTransition`, as we may want to eventually have a way to "skip" the current step. What do you think?

Reply:
Ah yes, this could be interesting. 

So I've placed this service in `xpack.core` (as it touches several abstractions and feels like a "core" functionality). This wouldn't allow us to reference `IndexLifecycleTransition` as that's `xpack.ilm`. 

However, given ILM is a big part in data tiers - the biggest bit in this transition service too- and it will be eaasier to  test the REST integration of this service (given the whole infrastructure for that is present in `xpack.ilm`) I now believe we should have the `MetadataMigrateToDataTiersRoutingService` and the corresponding REST API that's about to come in `xpack.ilm`.

What do you think?